src/tallies/filter_cell_instance.cpp

```diff
@@ -71,11 +76,30 @@ CellInstanceFilter::get_all_bins(const Particle& p, TallyEstimator estimator,
 {
   gsl::index index_cell = p.coord(p.n_coord() - 1).cell;
   gsl::index instance = p.cell_instance();
-  auto search = map_.find({index_cell, instance});
-  if (search != map_.end()) {
-    int index_bin = search->second;
-    match.bins_.push_back(index_bin);
-    match.weights_.push_back(1.0);
+
+  if (cells_.count(index_cell) > 0) {
+    auto search = map_.find({index_cell, instance});
+    if (search != map_.end()) {
+      int index_bin = search->second;
+      match.bins_.push_back(index_bin);
+      match.weights_.push_back(1.0);
+    }
+  }
+
+  if (material_cells_only_) return;
+
+  for (int i = 0; i < p.n_coord() - 1; i++) {
+    gsl::index index_cell = p.coord(i).cell;
+    // if this cell isn't used on the filter, move on
+    if (cells_.count(index_cell) == 0) continue;
+
+    // if this cell is used in the filter, check the instance as well
+    gsl::index instance = cell_instance_at_level(p, i);
```

Comment:
At each iteration of this loop we have O(i) work. If `p.n_coord()` is N, that means we'd have ∑(i, i=1..N) = O(N²) work, which seems a bit wasteful. Without using something over the top like C++20 coroutines, I'm not sure if there is a simple way to address this without repeating code. Only thing I can think of is to add two (optional) arguments to `cell_instance_at_level`, perhaps `starting_level` and `starting_instance`, and have it start at that level. Do you have a better idea?

Reply:
It's true that this would be the worst-case scenario, but that would only occur if every cell in `Particle::coord` were part of the cell instance filter, which in practice doesn't seem likely to me. The check above that the cell has instances in the filter should save us from computing the instance at each level of the geometry in most cases.

As far as the additional arguments for `cell_instance_at_level`, it's currently my understanding that to compute the correct instance for a given cell we have to start at the top of the geometry each time. Because each cell's `distribcell_index_` is different, it will have a different entry for each cell/lattice to look up as we move through the levels of `Particle::coord`. So unfortunately we can't use the instance information from the cell `a` at level `N` to inform the instance for cell `b` at level `N + 1` because we'd be looking up different information in levels `0` through `N`. Hopefully I'm understanding your intention correctly...

Reply:
Ok, I see what you mean. I agree that O(N²) is worst case and not very likely at all. The common case is probably O(N), which is as good as we can do for non-material filled cells.